core/domain/classifier_services.py

```diff
@@ -0,0 +1,534 @@
+        """
+        Returns a word's id if it exists, otherwise assigns
+        a new id to the word and returns it.
+        """
+        if word not in self._word_to_id:
+            self._word_to_id[word] = self._num_words
+            self._num_words += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """
+        Returns a label's id if it exists, otherwise assigns
+        a new id to the label and returns it.
+        """
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_label_name(self, l):
+        """
+        Returns a label's string name given its internal id.
+        If the id does not have a corresponding name, an exception is raised.
+        """
+        for label_name, label_id in self._label_to_id.iteritems():
+            if label_id == l:
+                return label_name
+
+        raise Exception('Label id %d does not exist.' % l)
+
+    def _get_doc_with_label_vector(self, d):
+        """
+        Given a doc id, return the doc and its label bit vector.
+        """
+        return self._w_dp[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """
+        Given a list of labels, return the label bit vector.
+        An empty label list is an alias for all labels being set. This is
+        useful because otherwise when a new label is introduce to the model,
+        none of the input data needs to be altered.
+        """
+        if len(labels) == 0:
+            return numpy.ones(self._num_labels)
+        label_vector = numpy.zeros(self._num_labels)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        label_vector[self._label_to_id[self._DEFAULT_LABEL]] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, val):
+        """
+        Updates counting matrices (ones that begin with _c) when a label
+        is assigned and unassigned to a word.
+        """
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _increment_counting_matrices(self, d, w, l):
+        """
+        Assign a label to a word in a doc.
+        """
+        self._update_counting_matrices(d, w, l, 1)
+
+    def _decrement_counting_matrices(self, d, w, l):
+        """
+        Unassign a label from a word in a doc.
+        """
+        self._update_counting_matrices(d, w, l, -1)
+
+    def _get_doc_ids(self, doc_ids):
+        """
+        Helper function for methods that iterate over doc ids.
+        If no doc ids are provided, then all doc ids are iterated over.
+        This simplifies logic closer to the client.
+        """
+        if doc_ids is None:
+            doc_ids = xrange(self._num_docs)
+        return doc_ids
+
+    def _assign_initial_labels_to_word_instances(self, doc_ids=None):
```

Comment:
Please note unresolved comment from first pass about documenting prerequisites for this method.